### PR TITLE
Fixes for Issues #833 and #834 (Schema updates for indexes)

### PR DESCRIPTION
**Pull request by iz-iznogood**
Implement Issue #833 by changing the SQL Server metadata query to return the whether an index is unique  or not. 
Implement issue #834 by extending the drop index logic to include the changes in columns - uniqueness so that the index is recreated properly

### CODE REVIEW

**Comment by iz-iznogood** [comment]
Please review and provide feedback

Thanks

File: src/schema-builder/RdbmsSchemaBuilder.ts [hunk 1]
```diff
@@ -357,7 +357,20 @@ export class RdbmsSchemaBuilder implements SchemaBuilder {
 
             // drop all indices that exist in the table, but does not exist in the given composite indices
             const dropQueries = tableSchema.indices
-                .filter(indexSchema => !metadata.indices.find(indexMetadata => indexMetadata.name === indexSchema.name))
+                .filter(indexSchema => {
+                    var metadataIndex = metadata.indices.find(indexMetadata => indexMetadata.name === indexSchema.name);
```
Comment (pleerock) [inline src/schema-builder/RdbmsSchemaBuilder.ts]: don't use `var` - use `const` instead. Also take a look, you have tslint errors

File: src/schema-builder/RdbmsSchemaBuilder.ts [hunk 2]
```diff
@@ -357,7 +357,20 @@ export class RdbmsSchemaBuilder implements SchemaBuilder {
...
+                    if (!metadataIndex)
+                        return true;
+                    if (metadataIndex.isUnique !== indexSchema.isUnique)
+                        return true;
+                    if (metadataIndex.columns.length !== indexSchema.columnNames.length)
+                        return true;
+                    for (let i = 0; i < metadataIndex.columns.length; ++i)
```
Comment (pleerock) [inline src/schema-builder/RdbmsSchemaBuilder.ts]: can you please make this code more functional? (without `for` loop)

**Comment by pleerock** [comment]
Please fix tslint issues, merge master into your branch (to get latest fixes to failing tests)

**Comment by pleerock** [comment]
We also need to make this to work on all platforms, not only mssql. Also would be great if we write tests for this functionality

File: src/schema-builder/RdbmsSchemaBuilder.ts (hunk 1 above)
Comment (ckalogeropoulos) [reply src/schema-builder/RdbmsSchemaBuilder.ts]: Done

File: src/schema-builder/RdbmsSchemaBuilder.ts (hunk 2 above)
Comment (ckalogeropoulos) [reply src/schema-builder/RdbmsSchemaBuilder.ts]: Done

**Comment by ckalogeropoulos** [comment]
Hi again, I fixed the review comments and I will send you a pull request soon.
However when testing it against MySQL, and Postgres I found several other issues in other areas.
In Postgres the indexes where updated fine but there is some issue with the sequences, the system is always trying to recreate them and fails. This is not related to this code
In MySQL the problems are worse, it tries to recreate the entire table in every run even there are no changes in the configuration. This is also unrelated to this code.
I could not test SQLite or Oracle since for SQLite I could not install the driver (some compile issues and I have no Oracle installed).
I will try to investigate the issues on Postgres and MySQL and fix them if possible.

**Comment by iz-iznogood** [comment]
Hi, please pull

**Comment by pleerock** [comment]
To make this PR complete we need to fix issues in mysql, postgres and sqlite. And add tests that confirms new fixes. Oracle can be skipped for now.

File: test/functional/database-schema/indices/indices-create-modify.ts [hunk 3]
```diff
@@ -0,0 +1,97 @@
...
+                expect(tableSchema!.indices.length).to.be.equal(1);
+                expect(tableSchema!.indices[0].name).to.be.equal("IDX_TEST"); 
+                expect(tableSchema!.indices[0].isUnique).to.be.false; 
+                expect(tableSchema!.indices[0].columnNames.length).to.be.equal(2); 
+                expect(tableSchema!.indices[0].columnNames[0]).to.be.equal("firstname"); 
+                expect(tableSchema!.indices[0].columnNames[1]).to.be.equal("lastname");
+    
+            };
+
+            const checkUniquenessFn = async function () {
```
Comment (pleerock) [inline test/functional/database-schema/indices/indices-create-modify.ts]: why did you make such structure? If you want separate blocks then better to create multiple `it` code blocks

Comment (iz-iznogood) [reply test/functional/database-schema/indices/indices-create-modify.ts]: Done

**Comment by iz-iznogood** [comment]
Hi, I have made some more fixes that I found while looking at #770.
This pull request is needed so that more fixes in other areas can come in (fix+test for #770).

Please review and merge if Ok. 
Thanks

**Comment by iz-iznogood** [comment]
Hi, this is getting a bit messy, I will provide a clean pull request with all the necessary commits.
Sorry for this

**Pull request closed without merging**